vowpalwabbit/core/src/reductions/cb/cb_explore_adf_graph_feedback.cc

```diff
@@ -0,0 +1,527 @@
+}
+
+void cb_explore_adf_graph_feedback::update_example_prediction(multi_ex& examples)
+{
+  auto& a_s = examples[0]->pred.a_s;
+  size_t num_actions = a_s.size();
+  arma::vec fhat(a_s.size());
+  for (auto& as : a_s) { fhat(as.action) = as.score; }
+
+  auto coord_gammafhat = set_initial_coordinates(fhat, _gamma);
+  arma::mat coordinates = std::get<0>(coord_gammafhat);
+  arma::vec gammafhat = std::get<1>(coord_gammafhat);
+
+  auto& graph_reduction_features =
+      examples[0]->ex_reduction_features.template get<VW::cb_graph_feedback::reduction_features>();
+  arma::sp_mat G = get_graph(graph_reduction_features, num_actions);
+
+  ConstrainedFunctionType f(gammafhat, G, _gamma);
+
+  ens::AugLagrangian optimizer;
+  optimizer.Optimize(f, coordinates);
+
+  // TODO json graph input
+  // TODO save_load the count and make gamma like squarecb gamma
+
+  arma::vec probs = get_probs_from_coordinates(coordinates, fhat, *_all);
+
+  // set the new probabilities in the example
+  for (auto& as : a_s) { as.score = probs(as.action); }
+  std::sort(
+      a_s.begin(), a_s.end(), [](const VW::action_score& a, const VW::action_score& b) { return a.score > b.score; });
+}
+
+template <bool is_learn>
+void cb_explore_adf_graph_feedback::predict_or_learn_impl(VW::LEARNER::learner& base, multi_ex& examples)
+{
+  if (is_learn)
+  {
+    _counter++;
+    base.learn(examples);
+    if (base.learn_returns_prediction) { update_example_prediction(examples); }
+  }
+  else
+  {
+    base.predict(examples);
+    update_example_prediction(examples);
+  }
+}
+
+void cb_explore_adf_graph_feedback::predict(VW::LEARNER::learner& base, multi_ex& examples)
+{
+  predict_or_learn_impl<false>(base, examples);
+}
+
+void cb_explore_adf_graph_feedback::learn(VW::LEARNER::learner& base, multi_ex& examples)
+{
+  predict_or_learn_impl<true>(base, examples);
+}
+
+void cb_explore_adf_graph_feedback::save_load(io_buf&, bool, bool) {}
+
+std::shared_ptr<VW::LEARNER::learner> VW::reductions::cb_explore_adf_graph_feedback_setup(
+    VW::setup_base_i& stack_builder)
+{
+  VW::config::options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+  using config::make_option;
+  bool cb_explore_adf_option = false;
+  bool graph_feedback = false;
+  float gamma = 0.f;
+
+  config::option_group_definition new_options(
+      "[Reduction] Experimental: Contextual Bandit Exploration with ADF with graph feedback");
+  new_options
+      .add(make_option("cb_explore_adf", cb_explore_adf_option)
+               .keep()
+               .necessary()
+               .help("Online explore-exploit for a contextual bandit problem with multiline action dependent features"))
+      .add(make_option("gamma", gamma).keep().default_value(10.f))
+      .add(make_option("graph_feedback", graph_feedback).necessary().keep().help("Graph feedback pdf").experimental());
+
+  auto enabled = options.add_parse_and_check_necessary(new_options) && graph_feedback;
```

Comment:
add_parse_and_check_necessary should check for all necessaries so the graph_feedback check is redundant